ta4j-core/src/main/java/org/ta4j/core/indicators/adx/ADXIndicator.java

```diff
@@ -37,26 +37,65 @@
  */
 public class ADXIndicator extends CachedIndicator<Num> {
 
+	private final DXIndicator dx;
     private final MMAIndicator averageDXIndicator;
     private final int diBarCount;
     private final int adxBarCount;
 
     public ADXIndicator(BarSeries series, int diBarCount, int adxBarCount) {
-        super(series);
-        this.diBarCount = diBarCount;
-        this.adxBarCount = adxBarCount;
-        this.averageDXIndicator = new MMAIndicator(new DXIndicator(series, diBarCount), adxBarCount);
-    }
+    	this(new ATRIndicator(series, diBarCount), adxBarCount);
+    } 
 
     public ADXIndicator(BarSeries series, int barCount) {
         this(series, barCount, barCount);
     }
+    
+    /*
+     * This constructor can be used in the unlikely event the user wants to share an ATR
+     * between ADX and another subsystem (Keltner channels, for example).
+     * 
+     * Users will normally use one of the bar series constructors,
+     * which will create all the necessary components.
+     * 
+     * Use getPlusDIIndicator and getMinusDIIndicator if you need access to those indicators.
+     * 
+     */
+    public ADXIndicator(ATRIndicator atr, int adxBarCount) {
+    	super(atr.getBarSeries());
+    	this.diBarCount = atr.getBarCount();
+        this.adxBarCount = adxBarCount;
+        this.dx = new DXIndicator(atr);
+        this.averageDXIndicator = new MMAIndicator(dx, adxBarCount);    	
+    }
 
+    /*
+     * This is obviously pure delegation.  
+     * ADX is caching the exact same values needlessly, I believe.
+     * ATR and 5-6 others do this too.
+     */
```

Comment:
Yes, this should be discussed and/or changed. But this comment should not be part of this lib, please remove.

Reply:
Yes..  those non-javadoc comment were intended for reviewers.  I will remove them.

I am not sure what you mean by adding to the change log.

Reply:
> Looking at the `PriceIndicator` I think it is over-engineered. Just for _calculating_ the close price you have an additional inheritance and an additional instance field holding the `priceFunction`.

Agree.  The price and volume classes could have been done as static fields or enums or something.  There are a few others too.  Typical price, median price... maybe as many as 10 classes could have been represented a simple lambdas like  b -> b.getClose() + b.getOpen() / 2... etc.  Coding Nums in lamdas is messy, but not cause for separate classes, IMO.  More classes to maintain...  more files...  oh well

Reply:
@team172011   

When you have a moment, I could use a little help,  I made a couple of rookie mistakes with commits yesterday and I would like to avoid doing it again.  I am new to this environment and the set of tools I am using may not be appropriate.
Here is what I have done:

- fork ta4j/ta4j into john../ta4j (6-7 days ago)
- clone to my local machine; prompted to install and use Github desktop, which is fine
- switch branches locally; my file system matches what I see on the server
- open eclipse; nice and seamless
- change code
- GitHub desktop sees changes and its easy to commit locally and push
- create pull request using one of these web pages

So far OK?  
 Questions
- how do I fetch from main to my repo;  my changelog file is out of date and I want to edit it
- how do I  automate this, so approved changes that go into main are pushed to my repo
- how do I resolve conflicts?  I must need a separate tool, as the web pages don't seem to give me a change to do anything but look; I will try to find a merge utility
- - maybe I can and should do all this through Eclipse; if you have tips for that I would appreciate it

thanks

Reply:
>maybe I can and should do all this through Eclipse; if you have tips for that I would appreciate it

Yes, definitly. You don't need to use Github Desktop. It's far easier to handle all of that within Eclipse. All your questions can be answered simply by using Git within Eclipse IDE. I recommend to read the official documentation: https://wiki.eclipse.org/EGit/User_Guide. The workflow is simple: 

1. Fetch Repo within Eclipse by using the View "Git Repository"
2. Import fetched Repo into Eclipse "Package Explorer"
3. make changes
4. Push or fetch changes from Eclipse (right click on "Package Explorer">"Team"-Menu)

Read https://wiki.eclipse.org/EGit/User_Guide or any guide you find by google with the keyword "_Eclipse EGit Tutorial_".

Reply:
Thank you very much.  I should have read some of this before asking questions.  I have made some progress.  I think the file set for this PR is what it should be.  2 files appear as conflicts.  I think they need to be approved and then merged.

Reply:
> I think they need to be approved and then merged.

No, conflicts says that **you** must resolve the conflict by yourself. Please edit

- CHANGELOG.md
- ta4j-examples/pom.xml 

so that the conflicts are resolved.

Reply:
I am afraid I am missing something.  I fetched the latest version of the changelog; I edited and committed it.  My changes were strictive additive.  I don't understand how this is causing a conflict.

The examples/pom.xml was changed after the 0.14 release, it seems.  I mistaken checked in my local pom with old version numbers.  I have fixed this and committed again.  

I'm afraid I do not know what git wants me to do

Reply:
Try to replace the whole pom.xml with our current version of the **`develop`**-branch.

Reply:
I have done that twice now.  Something is slightly different that I can't see.  I will look at it again tomorrow

Reply:
>Something is slightly different that I can't see.

Another alternative quick solution would be: 
1. close this PR
2. create a new branch "ADX Constructors" from the `develop`-branch. 
3. Adapt your changes (don't forget to include the changes in `changelog.md`). 
4. Submit PR.

Then it should work.